client/session.go

```diff
@@ -1625,74 +1661,163 @@ func (s *session) FetchBlocksFromPeers(
 
 func (s *session) streamBlocksMetadataFromPeers(
 	namespace ident.ID,
-	shard uint32,
-	peers []peer,
+	shardID uint32,
+	peers peers,
 	start, end time.Time,
+	consistencyLevel ReadConsistencyLevel,
 	metadataCh chan<- blocksMetadata,
 	resultOpts result.Options,
 	progress *streamFromPeersMetrics,
 	version FetchBlocksMetadataEndpointVersion,
 ) error {
 	var (
 		wg       sync.WaitGroup
-		errLock  sync.Mutex
-		errLen   int
-		pending  int64
-		multiErr = xerrors.NewMultiError()
+		errLock  sync.RWMutex
+		errors   = make(map[int]error)
+		setError = func(idx int, err error) {
+			errLock.Lock()
+			errors[idx] = err
+			errLock.Unlock()
+		}
+		getErrors = func() []error {
+			var result []error
+			errLock.RLock()
+			for _, err := range errors {
+				if err == nil {
+					continue
+				}
+				result = append(result, err)
+			}
+			errLock.RUnlock()
+			return result
+		}
+		abortError    error
+		setAbortError = func(err error) {
+			errLock.Lock()
+			abortError = err
+			errLock.Unlock()
+		}
+		getAbortError = func() error {
+			errLock.RLock()
+			result := abortError
+			errLock.RUnlock()
+			return result
+		}
+		pending   = int64(len(peers.peers))
+		majority  = atomic.LoadInt32(&s.majority)
+		enqueued  = int32(len(peers.peers))
+		responded int32
+		success   int32
 	)
+	if peers.selfExcluded {
+		state, err := peers.selfHostShardSet.ShardSet().LookupStateByID(shardID)
+		if err == nil && state == shard.Available {
+			// If we excluded ourselves from fetching, we basically treat ourselves
+			// as a successful peer response since we can bootstrap from ourselves
+			// just fine
+			enqueued++
+			success++
+		}
+	}
 
-	pending = int64(len(peers))
 	progress.metadataFetches.Update(float64(pending))
-	for _, peer := range peers {
+	for idx, peer := range peers.peers {
+		idx := idx
 		peer := peer
 
 		wg.Add(1)
 		go func() {
-			defer wg.Done()
-			var err error
-
-			switch version {
-			case FetchBlocksMetadataEndpointV1:
-				err = s.streamBlocksMetadataFromPeer(namespace, shard,
-					peer, start, end, metadataCh, progress)
-			case FetchBlocksMetadataEndpointV2:
-				err = s.streamBlocksMetadataFromPeerV2(namespace, shard,
-					peer, start, end, metadataCh, resultOpts, progress)
-			// Should never happen - we validate the version before this function is
-			// ever called
-			default:
-				err = errInvalidFetchBlocksMetadataVersion
+			defer func() {
+				// Success or error counts towards a response
+				atomic.AddInt32(&responded, 1)
+
+				// Decrement pending
+				progress.metadataFetches.Update(float64(atomic.AddInt64(&pending, -1)))
+
+				// Mark done
+				wg.Done()
+			}()
+
+			var currPageToken pageToken
```

Comment:
A comment here explaining that this starts off nil, but will be updated with the last seen page token when errors occur for the sake of resuming would help explain this code.

Actually maybe calling it `lastSeenPageToken` or something would also help

Reply:
Sure thing, I'll add a comment.